ibis/common/dispatch.py

```diff
@@ -106,3 +111,72 @@ def __call__(self, s: str, *args: Any, **kwargs: Any) -> Any:
     def __doc__(self) -> Any:
         # take the min to give the docstring of the last fallback function
         return min(self.priorities.items(), key=lambda x: x[1])[0].__doc__
+
+
+def lazy_singledispatch(func):
+    """A singledispatch implementation that supports lazily registering
+    implementations."""
+
+    lookup = {object: func}
+    abc_lookup = {}
+    lazy_lookup = defaultdict(dict)
+
+    def register(cls, func=None):
+        """Registers a new implementation for arguments of type `cls`"""
+
+        def inner(func):
+            if isinstance(cls, tuple):
+                for t in cls:
+                    register(t, func)
+            elif isinstance(cls, abc.ABCMeta):
+                abc_lookup[cls] = func
+            elif isinstance(cls, str):
+                module = cls.split(".", 1)[0]
+                lazy_lookup[module][cls] = func
+            else:
+                lookup[cls] = func
+            return func
+
+        return inner if func is None else inner(func)
+
+    def dispatch(cls):
```

Comment:
This all seems like it should be in its own library, but I guess we can always do that later!